drivers/sx127x: rework of implementations from #6645 and #6002

aabadie opened the pull request:
This PR is a rework of the initial SX1276 driver implementation from #6002 that was adapted by @jia200x in #6645.

The idea here is to make it ready for merge as fast as possible (targeting the coming release).

While adapting, I added basic support for SX1272, hence the driver name is changed to sx127x. I also kept the netdev adaptation.

I dropped the lorawan adaptation from #6645 because I think this should be provided as an external package (but maybe I'm wrong) and in a follow-up PR.

Note that I open this PR without heavy testing of the code (I don't have sx1276, only sx1272).

[This](http://www.st.com/content/st_com/en/products/evaluation-tools/product-evaluation-tools/stm32-nucleo-expansion-boards/p-nucleo-lrwan1.html) is the hardware I have.

PeterKietzmann commented on sys/include/net/netopt.h:
```diff
@@ -277,6 +300,8 @@ typedef enum {
     NETOPT_STATE_RESET,         /**< triggers a hardware reset. The resulting
                                  *   state of the network device is @ref NETOPT_STATE_IDLE */
+    NETOPT_STATE_STANDBY,       /**< standby mode. The devices is awake but
+                                 *   not listening to packets. */
```
Just out of interest, what do we need this mode for?

aabadie replied:
I think this is because the SX127x transceivers can be put in sleep mode and not listening to incoming packets. I guess this is one potential reason. @jia200x why did you add this ?

jia200x replied:
Hi, I'm back.
There's a standby mode in LoRa (is not the same as Sleep, the crystal oscillator is still running but doesn't receive packets). It's part of the RX-SINGLE workflow (after an rx timeout, the SX127X automatically goes back to standby, so shouldn't be ignored).

Cheers

PeterKietzmann commented on sys/include/net/netopt.h:
```diff
@@ -240,6 +240,29 @@ typedef enum {
+    NETOPT_LORA_MAX_PAYLOAD,                      /**< Maximum payload size */
+    NETOPT_LORA_RANDOM,                           /**< Random value */
+    NETOPT_LORA_TIME_ON_AIR,                      /**< Time on air */
+
```
I don't have a solution at hand but I'm wondering if we need/want all these options as generic netopts.

PeterKietzmann commented on tests/driver_sx127x/main.c:
```diff
@@ -0,0 +1,420 @@
+//        default:
+//            break;
+//    }
+//}
```
This is not so nice.

aabadie commented:
@PeterKietzmann thanks for the review, I'll try to address the comments. They all make a lot of sense.

FYI, the test application doesn't work as expected : I see some activity on the other side but no packet received yet.

aabadie commented:
@PeterKietzmann, I made small progress today : I can send and receive packet between 2 SX1272 modules. I need to cleanup the branch before pushing though.

dylad commented on drivers/include/sx127x.h:
```diff
@@ -0,0 +1,671 @@
+ * @param[in] dev                      The sx127x device descriptor
+ * @return the LoRa implicit mode
+ */
+bool sx12376_get_implicit_mode(sx127x_t *dev);
```
sx127x

dylad commented on tests/driver_sx127x/README.md:
```diff
@@ -0,0 +1,22 @@
+
+# Usage
+To setup bandwidth parameters, use lora_setup
+To change frequency, go to common.h
```
I think common.h is a legacy from #6002

dylad commented on tests/driver_sx127x/main.c:
```diff
@@ -0,0 +1,420 @@
+    return 0;
+}
+
+int config_channel_cmd(int argc, char **argv)
```
I think we need more verbose for this one
For example we need to set a random argv[1]  to set a channel

Cr0s commented:
Thanks for the effort to put these pieces together. Will check that on my SX1276 boards soon.

Cr0s commented on drivers/sx127x/include/sx127x_registers.h:
```diff
@@ -0,0 +1,1710 @@
+#define SX127X_RF_PREAMBLEDETECT_DETECTORTOL_2                       0x02
+#define SX127X_RF_PREAMBLEDETECT_DETECTORTOL_3                       0x03
+#define SX127X_RF_PREAMBLEDETECT_DETECTORTOL_4                       0x04
+#define SX127X_RF_PREAMBLEDETECT_DETECTORTOL_5          
```
Actually, most part of that registers macro is done by Semtech team in their LoRaMac's sx127x driver sources.

Cr0s commented on drivers/sx127x/include/sx127x_registers.h:
```diff
@@ -0,0 +1,1710 @@
+ */
+
+/**
+ * RegFifo
```
CamelCase is widely used in Semtech's reference design code from what that driver's sources and definition files was mostly inherited.

Cr0s commented on drivers/include/sx127x.h:
```diff
@@ -0,0 +1,671 @@
+#define SX127X_RADIO_WAKEUP_TIME                           (1000U)        /**< [us] */
+#define SX127X_RX_BUFFER_SIZE                              (256)          /**< RX buffer size */
+#define SX127X_RF_MID_BAND_THRESH                          (525000000UL)  /**< Mid-band threshold */
+#define SX127X_CHANNEL_HF                                  (868000000UL)  /**< HF channel */
```
SX127x has some kind of differentiation for so-called High and Low-frequency bands. There is a two different matching network layouts are required for both of the bands and two different outputs for each of them are implemented in a device. So we have to differentiate Low (below 868 MHz) and High modes in software by currently used channel frequency to correctly setup device's registers and use correct output that is dependent on a current channel frequency.

Cr0s commented on drivers/sx127x/sx127x_getset.c:
```diff
@@ -0,0 +1,599 @@
+
+            /* Note: When using LoRa modem only bandwidths 125, 250 and 500 kHz are supported. */
+            switch (dev->settings.lora.bandwidth) {
+                case 7: /* 125 kHz */
```
Good catch. There's a `enum` type `sx127x_lora_bandwidth_t` defined in `sx127x.h` and it has to be a `SX127X_BW_125_KHZ` member from it and for following cases as well.

aabadie commented on drivers/sx127x/include/sx127x_params.h:
```diff
@@ -0,0 +1,97 @@
+
+#ifndef SX127X_PARAM_DIO3
+#define SX127X_PARAM_DIO3                   GPIO_PIN(1, 4)       /* D5 */
+#endif
```
yes

aabadie commented on drivers/sx127x/include/sx127x_internal.h:
```diff
@@ -0,0 +1,140 @@
+ * @ingroup     drivers_sx127x
+ * @{
+ * @file
+ * @brief       Semtech SX1276 internal functions
```
fixed

aabadie commented on drivers/include/sx127x.h:
```diff
@@ -0,0 +1,671 @@
+    gpio_t dio3_pin;                                  /**< Interrupt line DIO3 */
+    gpio_t dio4_pin;                                  /**< Interrupt line DIO4 (not used) */
+    gpio_t dio5_pin;                                  /**< Interrupt line DIO5 (not used) */
+} sx127x_params_t;
```
done

aabadie commented on drivers/sx127x/include/sx127x_registers.h:
```diff
@@ -0,0 +1,1710 @@
+ * SX127X Internal registers Addresses
+ */
+
+/** Common settings */
```
should be fixed

aabadie commented on drivers/sx127x/sx127x.c:
```diff
@@ -0,0 +1,478 @@
+
+    dev->irq |= SX127X_IRQ_DIO5;
+    sx127x_isr((netdev_t*) dev);
+}
```
I don't known how to do that

aabadie commented:
@PeterKietzmann, @Cr0s, @dylad thanks for your comments.
I just pushed a new branch (force push sorry) were I tried to address the required changes. At least the static tests pass locally and I can exchange messages between 2 modules.

I also refactored a bit the test application.

aabadie commented:
@PeterKietzmann @dylad did you find time for some tests ?

dylad commented:
@aabadie I plan to do some tests tomorrow I'll let you know.

BTW, I would like to know if someone has a SX1272RF1BAS board ? I would like to check something...

PeterKietzmann commented:
Unfortunately not. I have 2 x [SX1276MB1MAS](http://www.mouser.de/Search/ProductDetail.aspx?R=SX1276MB1MASvirtualkey57390000virtualkey947-SX1276MB1MAS) and 2 x [P-NUCLEO-LRWAN1](http://www.mouser.de/Search/ProductDetail.aspx?R=P-NUCLEO-LRWAN1virtualkey51120000virtualkey511-P-NUCLEO-LRWAN1). No SX1272RF1BAS, sorry.

aabadie commented:
rebased.

@PeterKietzmann, do you plan to port RIOT to the P-NUCLEO-LRWAN1 ? It will be great and not so difficult I think (stm32l0 are supported by RIOT).

dylad commented:
@aabadie 
I did some tests today with my boards but I have (maybe ?) a problem with my SX1272RF1 board. There are old and it seems their revision is not 0x22 (from reg 0x42) because there are engineering sample released long ago. So it supposed to work but without any warranty.

I tried to exchanges messages with one SX1276-based board and one of my old SX1272 and It doesn't work. (Maybe because my samples are old ?)
But, I found out SX127X_REG_LR_MODEMCONFIG 1&2 are different for SX1272 and SX1276

(SX127X_REG_LR_MODEMCONFIG3 exists for SX1276 but is reserved for SX1272). So several functions cannot work for SX1272 chip like `sx127x_set_bandwidth`.

It may have others registers differents but I didn't have much time to investigate further.

PeterKietzmann commented:
@aabadie I'm a bit confused. Actually I thought that you are working with that shield (based on your comment [here](https://github.com/RIOT-OS/RIOT/pull/6002#issuecomment-287347352)).

aabadie commented:
@PeterKietzmann, sorry, you are right, those are the one I have... I was thinking of this board: [B-L072Z-LRWAN1](http://www.st.com/en/evaluation-tools/b-l072z-lrwan1.html). This one is very interesting for porting RIOT on.

aabadie commented:
@dylad thanks for spotting those differences. I'll look into that.
Testing the communication between the 2 different types (SX1272 and SX1276) is also a good test for interop.

dylad commented on drivers/sx127x/sx127x.c:
```diff
@@ -0,0 +1,450 @@
+    gpio_init_int(dev->params.dio0_pin, GPIO_IN, GPIO_RISING, sx127x_on_dio0_isr, dev);
+    gpio_init_int(dev->params.dio1_pin, GPIO_IN, GPIO_RISING, sx127x_on_dio1_isr, dev);
+    gpio_init_int(dev->params.dio2_pin, GPIO_IN, GPIO_RISING, sx127x_on_dio2_isr, dev);
+    gpio_init_int(dev->params.dio3_pin, GPIO_IN, GPIO_RISING, sx127x_on_dio3_isr, dev);
```
@aabadie Could you add some verbose if those pins are successfully init or not please ?
I had some surprises with this.

aabadie replied:
Just pushed a commit @dylad. With ST MCU, the `gpio_init_int` always return 0 so it won't be helpful. I guess your are trying with a SAMD21 ? Also don't forget to set `ENABLE_DEBUG (1)` at the beginning of the file.

aabadie replied:
Out of curiosity, what are the surprises you had ?

dylad replied:
Many thanks @aabadie  !
I used a SAML21 and I set PA08 as DIO0 without looking and of course..., I just found out PA08 can't be used as external interrupt.
I'll retest tomorrow (if possible) with another pin.

dylad commented on tests/driver_sx127x/main.c:
```diff
@@ -0,0 +1,368 @@
+
+    if (argc < 4) {
+        puts("usage: setup "
+             "<bandwidth (125, 250, 512)> "
```
512 -> 500

aabadie replied:
fixed

aabadie commented:
@dylad, I pushed a commit that should fix the problem you mention about the `SX127X_REG_LR_MODEMCONFIG` registers.
According to the datasheets, number 1 & 2 have the same addresses for both type of modules (see page 91 of both [SX1272](http://www.semtech.com/images/datasheet/sx1272.pdf) and [SX1276](http://www.semtech.com/images/datasheet/sx1276.pdf) datasheets.
But you are right that number 3 is reserved in SX1272 so it cannot be used to optimize the data rate in `set_bandwidth`. I updated the implementation to reflect what is described in the datasheet.

Thanks for spotting this :)

dylad commented on drivers/sx127x/sx127x_getset.c:
```diff
@@ -0,0 +1,623 @@
+    tmp |= bandwidth << 4;
+
+    dev->settings.lora.bandwidth = bandwidth;
+    sx127x_reg_write(dev, SX127X_REG_LR_MODEMCONFIG1, tmp);
```
There is still a problem with SX1272, need rewrite.

dylad commented on drivers/sx127x/sx127x_getset.c:
```diff
@@ -0,0 +1,623 @@
+bool sx127x_get_crc(sx127x_t *dev)
+{
+    return (sx127x_reg_read(dev, SX127X_REG_LR_MODEMCONFIG2) &
+            SX127X_RF_LORA_MODEMCONFIG2_RXPAYLOADCRC_MASK);
```
Need rewrite for SX1272

dylad commented on drivers/sx127x/sx127x_getset.c:
```diff
@@ -0,0 +1,623 @@
+    uint8_t tmp = sx127x_reg_read(dev, SX127X_REG_LR_MODEMCONFIG1);
+    tmp &= SX127X_RF_LORA_MODEMCONFIG1_IMPLICITHEADER_MASK;
+    tmp |= implicit;
+    sx127x_reg_write(dev, SX127X_REG_LR_MODEMCONFIG1, tmp);
```
need rewrite for SX1272

dylad commented on drivers/sx127x/sx127x_internal.c:
```diff
@@ -0,0 +1,249 @@
+    sx127x_set_op_mode(dev, SX127X_RF_OPMODE_STANDBY);
+    sx127x_reg_write(dev, SX127X_REG_LR_MODEMCONFIG1, SX127X_RNG_REG_MODEM_CONFIG1);
+    sx127x_reg_write(dev, SX127X_REG_LR_MODEMCONFIG1, SX127X_RNG_REG_MODEM_CONFIG2);
+
```
SX127X_REG_LR_MODEMCONFIG2 otherwise you rewrite the same register.
Watch out : SX127X_RNG_REG_MODEM_CONFIG (1&2) are different for SX1272 & SX1276

dylad commented on drivers/sx127x/sx127x.c:
```diff
@@ -0,0 +1,461 @@
+    xtimer_usleep(1000);
+
+    /* Put reset pin in High-Z */
+    gpio_init(dev->params.reset_pin, GPIO_OD);
```
Sorry to ask it again, but could you also add some verbose here ? SAML21 doesn't support GPIO_OD and it took me some time to figured out why my SX1272 was always return 0x00 (It was stuck in reset state).

BTW, should we think of an alternate solution for MCU which don't have OD output ? like set the pin as input ?
I know it is ugly but I didn't have any better idea...

aabadie replied:
Looking at the code, it should be configured as input.

aabadie replied:
after discussing IRL with @haukepetersen, configuring as input seems reasonable. Can you test that it works with your SAML21 ?

dylad replied:
I already tested it. It works like a charm !

aabadie commented:
> SX1272 has 2 bits for set bandwidth in CONFIGMODEM1 but SX1276 has 4 bits

After looking more closely in the datasheet, you are right and this is problem. Thanks again for spotting this. I'm trying to find a solution.

aabadie commented:
by the way, the changes with bandwidth, coding rate, CRC, etc is fine for me. I could play with different configurations with success (before it was obviously not the case). So thanks a lot for your help @dylad !

dylad commented:
Regarding `SX127X_RNG_REG_MODEM_CONFIG` I also have no clue...
I took a quick look at SX1272Lib & SX1276Lib from mbed and the random function is a bit diferent...
[SX1272Lib](https://developer.mbed.org/teams/Semtech/code/SX1272Lib/file/69d5976b504d/sx1272/sx1272.cpp)
[SX1276Lib](https://developer.mbed.org/teams/Semtech/code/SX1276Lib/file/3778e6204cc1/sx1276/sx1276.cpp)
@Cr0s @jia200x  any idea ? (Maybe should we remove it ? ...)

dylad commented on drivers/sx127x/sx127x_getset.c:
```diff
@@ -0,0 +1,717 @@
+            break;
+        case SX127X_BW_500_KHZ:
+            config1_reg |=  SX1276_RF_LORA_MODEMCONFIG1_BW_500_KHZ << 4;
+            break;
```
Either remote shifting or change SX1276_RF_LORA_MODEMCONFIG1_BW_XXX_KHZ for SX1276 device
For example, in the current state SX1276_RF_LORA_MODEMCONFIG1_BW_125_KHZ is 0x70. So 0x70 << 4 We lost our bits. 
SX1272 seems fine.

aabadie replied:
ok, I see, will fix

aabadie replied:
fixed

dylad commented:
@aabadie changes looks great except for the comment about bandwidth for SX1276.
I was also able to share messages between one SX1272 and one SX1276 in both direction.
Change coding rate and spreading factor works perfectly.
But change in bandwidth is broken for SX1276 so I still need to test when you fix it.
Great work !

dylad commented on drivers/sx127x/sx127x_getset.c:
```diff
@@ -0,0 +1,717 @@
+        config1_reg &= SX1276_RF_LORA_MODEMCONFIG1_BW_MASK;
+        switch (dev->settings.lora.bandwidth) {
+        case SX127X_BW_125_KHZ:
+            config1_reg |= (SX1276_RF_LORA_MODEMCONFIG1_BW_125_KHZ << 4);
```
Coding style comment: Why parenthesis on this one and not the others ?

aabadie replied:
I forgot to change this

aabadie commented:
> I was also able to share messages between one SX1272 and one SX1276 in both direction.

This makes my day. Thanks @dylad :)

dylad commented on drivers/sx127x/include/sx127x_registers.h:
```diff
@@ -0,0 +1,1732 @@
+#define SX1276_RF_LORA_MODEMCONFIG1_BW_250_KHZ                       0x80
+#define SX1276_RF_LORA_MODEMCONFIG1_BW_500_KHZ                       0x90
+
+#define SX1272_RF_LORA_MODEMCONFIG1_CODINGRATE_MASK                  0xC2
```
0xC2 -> 0xC7 otherwise this mask breaks CONFIG1 register for SX1272

aabadie replied:
fixed

dylad commented on drivers/sx127x/include/sx127x_registers.h:
```diff
@@ -0,0 +1,1732 @@
+#define SX127X_RF_LORA_MODEMCONFIG1_CODINGRATE_4_5                   0x02
+#define SX127X_RF_LORA_MODEMCONFIG1_CODINGRATE_4_6                   0x04 /**< Default */
+#define SX127X_RF_LORA_MODEMCONFIG1_CODINGRATE_4_7                   0x06
+#define SX127X_RF_LORA_MODEMCONFIG1_CODINGRATE_4_8                   0x08
```
I think you should change to : 
```
#define SX127X_RF_LORA_MODEMCONFIG1_CODINGRATE_4_5                   0x01
#define SX127X_RF_LORA_MODEMCONFIG1_CODINGRATE_4_6                   0x02
#define SX127X_RF_LORA_MODEMCONFIG1_CODINGRATE_4_7                   0x03
#define SX127X_RF_LORA_MODEMCONFIG1_CODINGRATE_4_8                   0x04
```

aabadie replied:
indeed

aabadie replied:
but in fact it's not used in the code

aabadie commented:
> Regarding SX127X_RNG_REG_MODEM_CONFIG I also have no clue...
I took a quick look at SX1272Lib & SX1276Lib from mbed and the random function is a bit diferent...

@dylad, indeed the functions are different, the `SX127X_RNG_REG_MODEM_CONFIG` is not used. I removed the lines and the random function still works. I pushed an update accordingly.

adjih commented:
Tested it and base functionality (sending one packet with the test example parameter) works well, so it is an ACK. 

I have no opinion on the reporting of the LQI/RSSI/noise/...  but it seems to be not so straighforward.

jia200x commented:
@aabadie no idea what the SX127X_RNG_REG_MODEM_CONFIG does. But as I see RNG is still working ;).

I'm into the SX1276 bandwidth problem now.

@adjih the current code should report the SNR and RSSI via netdev2->recv.

Cheers!

dylad commented:
@jia200x Could you also check if spreading factor/ coding rate are set correctly within register? (I think there are still some minors issues with them) I'll do more tests tomorrow.

jia200x commented:
@dylad Ok, I will check it

jia200x commented:
Spreading factor is OK. Coding rate is not being set correctly in SX1276. It might be the shell command, I'm checking.

jia200x commented:
[Found the problem].

when 'cr' is 5, the "lora_cr" variable is 0. In the sx127x_set_coding_rate function, this causes the 3-1 bytes of RegModemConfig1 (0x1D) to be as 0. For CR5 should be set to 001. So, all CR are being set as the previous one (CR6 => CR5, CR7 => CR6, etc). 

[Found the problem]: https://github.com/RIOT-OS/RIOT/pull/6797/files#diff-50e488c3bf4bb01701eb814803bd0ef2R105

aabadie commented:
nice catch @jia200x, I'll fix it.

aabadie commented:
> the current code should report the SNR and RSSI via netdev2->recv

I pushed a commit that slightly adapt SNR to use signed values since LoRa can work with negative SNR apparently. The LQI is ignored.

adjih commented:
@jia200x  / @aabadie ->  LoRa can indeed decode packets below the noise floor (so SNR < 0 dB) ;   but then, related to issue #6887 , consider that RSSI can be below - 128 (dBm)

jia200x commented:
@adjih yes, for sure. We will take that into consideration.

@dylad I'm able to change bandwidth in sx1276 (didn't test it, but I can see it in registers). Is not working for you?

aabadie commented:
@jia200x, just a commit that makes a bit of cleanup.
The BW, SF and CR values update should work ok now (I normally fixed that last week). The RSSI value, that is potentially negative when returned in dBm, is a more general problem in netdev and I don't think it should be solved in this PR.

I don't see any remaining issues here, so it would be great to have a more in depth code review (if possible) so we can merge this one soon.
The next steps I see to reach LoRa support are:
* add `sx127x` in `auto_init`
* add CMAC implementation from #6645
* add LoRaMac adaptation

What do you think ?

dylad commented:
@jia200x As I remember, I could not exchange messages between one SX1272 and SX1276 (or SX1276 and a gateway based on SX1301/SX1252) with a bandwidth > 125kHz.
I don't know if I'll have enough time this week to do further tests (+ a review). I'll try.

aabadie commented:
@dylad, can you try again the exchange messages between SX1272 and SX1276 ? The bandwidth problem should be fixed.

dylad commented:
@aabadie could you confirm you can change bandwidth with SX1272 (and check in register 0x1D if it's ok)
I'm currently trying, but there is no change in 0x1D when I set a new bandwidth. This is weird...

dylad commented on drivers/sx127x/sx127x_getset.c:
```diff
@@ -0,0 +1,723 @@
+
+        sx127x_reg_write(dev, SX127X_REG_LR_MODEMCONFIG1,
+                         (sx127x_reg_read(dev, SX127X_REG_LR_MODEMCONFIG1) &
+                          SX127X_RF_LORA_MODEMCONFIG1_LOWDATARATEOPTIMIZE_MASK) |
```
Aaaaah, this one gives me a huge headache....
Currently `SX127X_RF_LORA_MODEMCONFIG1_LOWDATARATEOPTIMIZE_MASK` is 0x01 so this call clear all the bits instead of protect them. This mask must be 0xFE to do the correct job.
This bug prevents SX1272 to set a bandwidth superior to 125 kHz...
You can check it using register get 0x1D shell command before and after setup a new bandwidth.

aabadie replied:
Thanks for the in-depth test @dylad. I'll test again tomorrow.

aabadie replied:
That was it. Thanks, I pushed the fix, the register is now correctly updated.

jia200x commented:
@aabadie I agree with the next steps for LoRaWAN support :)

There are some things still to be done in the LoRaWAN code, but shouldn't be hard.
Cheers

dylad commented:
It would be great if RIOT has LoRaWAN support.
Class A & C should not be hard. But Class B will be more challenging.

jia200x commented:
@dylad Check #6645. It's a full LoRaWAN implementation (Class A & C).

dylad commented:
@jia200x I know I watched your PR since the very beginning ;)
But we have to test, review... It will take some times.

aabadie commented:
@jia200x, @dylad, as I said, the best is open a new PR providing only CMAC support (this can be taken from #6645) and then open a second PR based on #6797 and the CMAC PR just providing LoRaMAC.

aabadie commented:
@jia200x @dylad, I'll open a PR for CMAC in a moment

aabadie commented:
ok, it seems that opening a new PR for CMAC is not needed since there's already one: #6109 :)

dylad commented:
I took some homeworks with me for the weekend.
I'm now able to exchange messages between one SX1272 and one SX1276 with all bandwidth, SF and coding rate !
But there is still a problem !
When I set both chip to `setup 250 12 5` (Bandwidth 250 kHz, Spreading factor 12, coding rate 4/5)
I cannot exchange messages. I checked the registers, the LowDataRateOptimize bit is set to 1 for SX1276, and 0 for SX1272. I then try to set it to 1 manually. Once both chips have LowDataRateOptimize set to 1, I was able to exchange messages. 
Semtech datasheet said : 

> Note that both the transmitter and the receiver must have the same setting for 
LowDataRateOptimize

Otherwise, It works like a charm !

aabadie commented:
@dylad you are right about the configuration required on both side.
Looking at the datasheet it's not clear to me what's the correct configuration about data rate optimization:
* for sx1276 (section 4.1.1.6), the lowdatarateoptimize should be set when "preamble duration exceeds 16ms"
* for sx1272, it has to be set when BW is 125kHz and SF is either 11 or 12

If you say that it works even when you force it, I'll update the code to be the same for both (taking sx1272 as reference)

dylad commented:
I'd rather think we should take inspiration from Semtech again. As I saw, LowDataRateOptimize is handle by SX1272 & SX1276 in the same way.
[SX1272Lib](https://developer.mbed.org/teams/Semtech/code/SX1272Lib/file/69d5976b504d/sx1272/sx1272.cpp)
[SX1276Lib](https://developer.mbed.org/teams/Semtech/code/SX1276Lib/file/3778e6204cc1/sx1276/sx1276.cpp)
What do you think ?

aabadie commented:
> What do you think ?

Yes, indeed it's done the same way. I pushed an update, should be ok now.

aabadie commented:
BTW, the link to the libraries are interesting: the FSK modem is supported there. Porting this to RIOT shouldn't be too hard and would provide [802.15.4g](http://www.keysight.com/upload/cmc_upload/All/Slide_IOT_Part_2.pdf?&cc=FR&lc=fre) radio.

dylad commented:
Why not, but do you plan to add support for FSK in the current PR ?

aabadie commented:
> do you plan to add support for FSK in the current PR ?

nope, let's first merge this one and move forward on full LoRawan support. FSK can/should be added in a follow-up.

jia200x commented:
@travisgriggs The AT86RF2xx is not related at all. That part was based in the header file of the AT86RF (copy & pasta).
To be fixed.

Cheers!

aabadie commented:
@travisgriggs, thanks for spotting this. It's fixed now.

dylad commented on drivers/sx127x/sx127x_getset.c:
```diff
@@ -0,0 +1,735 @@
+}
+
+static inline uint8_t sx127x_get_pa_select(uint32_t channel)
+{
```
@aabadie @jia200x 
Why PABOOST is linked to  channel frequency ? Shouldn't we let the user choose ?

aabadie replied:
Looking at the datasheets, this is something related to SX1276. SX1272 doesn't work the same way. This has to be reworked.

aabadie replied:
The SX1272 transmitter mode was broken because of this change. I reverted the PA_SELECT to RFO for SX1272. For SX176 it's link to the channel used (use PABOOST in 868 MHz).
Surprisingly this is not the case [here](https://github.com/Lora-net/LoRaMac-node/blob/master/src/boards/LoRaMote/sx1272-board.c#L156), and their code shouldn't work.

dylad replied:
Which hardware are you using with your SX1272 ?
I just looked at [SX1272RF1BAS schematic](http://www.semtech.com/images/datasheet/sx1272ska_userguide.pdf), RFO is unconnected, you can only transmit using PABOOST with this module. This seems to be hardware-related.

aabadie replied:
@dylad I have [this one](https://developer.mbed.org/media/uploads/GregCr/sx1272mb2xas_e364v01a_sch.pdf) and RFO is connected.

aabadie replied:
The module I have is described [here](https://developer.mbed.org/components/SX1272MB2xAS/)

dylad replied:
Seems there are several configurations...
But in the current situation, your module must works but not mine. Maybe we should diverge from Semtech for this case ?

aabadie replied:
My opinion is that mbed modules are designed to be used with arduino compatible boards (nucleos for instance). I would prefer to stick to them by default. But I can provide a way to configure the driver (using macros) if you want : something like `SX1272_USE_PABOOST`. What do you think ?

dylad replied:
`mbed modules are designed to be used with arduino compatible boards` Yes
But we should also let the possibility to the user to use custom board/sx127x couple with their own hardware specification.
If there is a way to configure SX1272 to use PA_BOOST without hacking the getset.c file I'm okay :)

aabadie replied:
> But we should also let the possibility to the user to use custom board/sx127x couple with their own hardware specification.

Agreed

> If there is a way to configure SX1272 to use PA_BOOST without hacking the getset.c file I'm okay :)

I think the global macro I suggested + some minor changes related in getset.c should make it.
I'll push an update soon. Still working on @miri64 required changes for the moment.

dylad commented on drivers/sx127x/sx127x_getset.c:
```diff
@@ -0,0 +1,735 @@
+            xtimer_set(&(dev->_internal.rx_timeout_timer),
+                       dev->settings.window_timeout);
+        }
+        sx127x_set_op_mode(dev, SX127X_RF_LORA_OPMODE_RECEIVER_SINGLE);
```
Cannot apply this line to SX1272

dylad commented on drivers/sx127x/sx127x_getset.c:
```diff
@@ -0,0 +1,735 @@
+                    (8 * pkt_len - 4 * dev->settings.lora.sf + 28
+                     + 16 * dev->settings.lora.crc_on
+                     - (!dev->settings.lora.implicit_header_mode ? 20 : 0))
+                    / (double) (4 * dev->settings.lora.sf
```
There are several difference between this calculation and SX1272Lib & SX1276Lib

aabadie replied:
I could only find one (compared with [SX1276](https://github.com/Lora-net/LoRaMac-node/blob/master/src/radio/sx1276/sx1276.c) and [SX1272](https://github.com/Lora-net/LoRaMac-node/blob/master/src/radio/sx1272/sx1272.c))

dylad commented on drivers/sx127x/sx127x_getset.c:
```diff
@@ -0,0 +1,735 @@
+            double t_on_air = t_preamble + t_payload;
+
+            /* return seconds */
+            air_time = floor(t_on_air * 1e6 + 0.999);
```
Semtech used 1e3 instead of 1e6 (possible error ?)

aabadie replied:
this one exactly: should be 1e3

dylad commented on drivers/sx127x/sx127x_getset.c:
```diff
@@ -0,0 +1,735 @@
+    if (op_mode != op_mode_prev) {
+        /* Replace previous mode value and setup new mode value */
+        sx127x_reg_write(dev, SX127X_REG_OPMODE,
+                         (op_mode_prev & SX127X_RF_OPMODE_MASK) | op_mode);
```
op_mode_prev on this line is useless to me.

dylad commented on drivers/sx127x/sx127x_getset.c:
```diff
@@ -0,0 +1,735 @@
+    uint8_t pa_dac = sx127x_reg_read(dev, SX127X_REG_PADAC);
+
+    pa_config = ((pa_config & SX127X_RF_PACONFIG_PASELECT_MASK) |
+                 sx127x_get_pa_select(dev->settings.channel) << 7);
```
remove << 7. 
`sx127x_get_pa_select` returns either 0x80 or 0x00 so << is a problem.

dylad commented on drivers/sx127x/sx127x_getset.c:
```diff
@@ -0,0 +1,735 @@
+    pa_config = ((pa_config & SX127X_RF_PACONFIG_PASELECT_MASK) |
+                 sx127x_get_pa_select(dev->settings.channel) << 7);
+    /* max power is 14dBm */
+    pa_config = (pa_config & SX127X_RF_PACONFIG_MAX_POWER_MASK) | (0x05 << 4);
```
There is no max power for SX1272.

dylad commented on drivers/sx127x/sx127x_getset.c:
```diff
@@ -0,0 +1,735 @@
+    }
+
+    sx127x_reg_write(dev, SX127X_REG_PACONFIG, pa_config);
+    sx127x_reg_write(dev, SX127X_REG_PADAC, pa_dac);
```
`SX127X_REG_PADAC` this register doesn't have the same address for SX1272 & SX1276. 
In the current state, it only works for SX1276

dylad commented on drivers/sx127x/sx127x_getset.c:
```diff
@@ -0,0 +1,735 @@
+                power = 20;
+            }
+
+            pa_config = ((pa_config & SX127X_RF_PACONFIG_OUTPUTPOWER_MASK) |
```
Need to rewrite this (and others `pa_config` within this function) to match SX1272 register which is different from SX1276

aabadie replied:
Just had a look at the datasheets of both devices and the PaConfig register is the same. Only Max power is unused for sx1272. So I don't think there's need for rewrite or maybe I'm missing something

dylad replied:
If it's not harmful, then we can leave it as it.

dylad commented on drivers/sx127x/sx127x_getset.c:
```diff
@@ -0,0 +1,735 @@
+    dev->settings.lora.iq_inverted = iq_invert;
+
+    uint8_t invertiq_reg = sx127x_reg_read(dev, SX127X_REG_LR_INVERTIQ);
+    invertiq_reg &= SX127X_RF_LORA_INVERTIQ_RX_MASK;
```
Is it really usefull ? this bit is all alone in the register. I think we can optimize here.

dylad commented on drivers/sx127x/sx127x_getset.c:
```diff
@@ -0,0 +1,735 @@
+                             SX127X_RF_LORA_OPMODE_LONGRANGEMODE_ON);
+
+            sx127x_reg_write(dev, SX127X_REG_DIOMAPPING1, 0x00);
+            sx127x_reg_write(dev, SX127X_REG_DIOMAPPING2, 0x10); /* DIO5=ClkOut */
```
Why 0x10 ? Semtech set this register to 0x00.

aabadie replied:
Indeed, good point

dylad commented on drivers/sx127x/sx127x.c:
```diff
@@ -0,0 +1,491 @@
+    /* Copy LoRa configuration into device structure */
+    memcpy(&dev->settings.lora, &lora_settings, sizeof(sx127x_lora_settings_t));
+
+    sx127x_set_modem(dev, SX127X_MODEM_LORA);
```
`sx127x_set_modem ` is called several time during init. We can optimize this.

dylad commented on drivers/sx127x/sx127x.c:
```diff
@@ -0,0 +1,491 @@
+    sx127x_set_hop_period(dev, dev->settings.lora.freq_hop_period);
+    sx127x_set_power(dev, dev->settings.lora.power);
+
+    sx127x_set_channel(dev, LORA_ISM_FREQUENCY);
```
There is several call for this during init. We can also optimize.

dylad commented on drivers/sx127x/sx127x.c:
```diff
@@ -0,0 +1,491 @@
+    sx127x_set_op_mode(dev, SX127X_RF_OPMODE_STANDBY);
+
+    /* Set radio in continuous reception */
+    sx127x_set_op_mode(dev, SX127X_RF_OPMODE_RECEIVER);
```
Just a question here : Is it normal to have two call of `sx127x_set_op_mode` in a row ?

aabadie replied:
apparently not (comparing to [LoRaMac-node/sx1276.c](https://github.com/Lora-net/LoRaMac-node/blob/master/src/radio/sx1276/sx1276.c#L306)

dylad commented on drivers/sx127x/sx127x_getset.c:
```diff
@@ -0,0 +1,735 @@
+            }
+            else {
+                sx127x_reg_write(dev, SX127X_REG_LR_DETECTOPTIMIZE,
+                                 sx127x_reg_read(dev, SX127X_REG_LR_DETECTOPTIMIZE) | 0x80);
```
This is black magic ! This bit is not documented in both SX1272 and SX1276.
But Semtech set this bit (for SX1276 only)
What it is this bit ?

aabadie replied:
I don't know either

dylad commented on drivers/sx127x/sx127x_getset.c:
```diff
@@ -0,0 +1,735 @@
+            /* FSK not supported yet */
+            return;
+
+        case SX127X_MODEM_LORA:
```
About this case
Won't work with SX1272 according to SX1272Lib. Need to be rewrite for SX1272 chip.

aabadie replied:
the only different is that there's no check on bandwidth values for sx1272

aabadie commented:
@dylad, your in-depth review is great, it really helps. I'll try to address the issues asap !

dylad commented:
@aabadie Any news about this PR ?

aabadie commented:
> Any news about this PR ?

Well nothing during the last 3 weeks but this is getting upper on my stack. I wish I can work on it in the coming week.

PeterKietzmann commented on cpu/cc2538/radio/cc2538_rf_getset.c:
```diff
@@ -230,6 +230,8 @@ void cc2538_set_state(cc2538_rf_t *dev, netopt_state_t state)
             dev->state = NETOPT_STATE_IDLE;
             break;
+        default:
+            break;
```
How is this related to this PR? It is picky, sure, but could you make a separate PR out of it in case it is unrelated?

aabadie replied:
Because this PR introduces a new netopt state for LoRA which is not handled by this switch. Thus the build for cc2538 was raising `warning: enumeration value '[...]'not handled in switch [-Wswitch]` => error.
I can open a dedicated PR if you prefer.

PeterKietzmann replied:
Ah now I see. No, leave it as is.

aabadie commented on drivers/sx127x/sx127x_getset.c:
```diff
@@ -0,0 +1,735 @@
+                sx127x_reg_write(dev, SX127X_REG_LR_INVERTIQ2, SX127X_RF_LORA_INVERTIQ2_OFF);
+            }
+
+            /* ERRATA 2.3 - Receiver Spurious Reception of a LoRa Signal */
```
This block is not needed for sx1272

aabadie commented:
@dylad, I addressed your comments. Sorry for taking so long, I couldn't find enough time to efficiently work on this.

By the way, I changed the way we distinguish sx1272 and sx1276: now sx1272 and sx1276 are defined as pseudomodules depending on sx127x and I use preprocessor macros for device specific code. This should save memory since we don't need to use switch/case everywhere anymore.

Before merging, I would also like to refactor the test application, finalize the gnrc_netdev adaption and configure the auto initialization. This way it will easier (I think) to integrate the Loramac part.

Will you find some time to have a look before feature freeze (in 2 weeks) ? The objective is to have this work in the coming release.

dylad commented:
> This should save memory since we don't need to use switch/case everywhere anymore


That's definitely the best solution

I'll try to review one more time + test ASAP.I hope this PR will be merged for the coming release !

aabadie commented:
ping @dylad, could you find some time for testing/reviewing ?

dylad commented:
I'll try to finish my review for tomorrow evening. But I have no idea if I'll be able to fully test this PR using one SX1272 and one SX1276. I'll be very busy at work and at home for the whole summer...

aabadie commented:
That's fine for the review. I tested with SX1276 and SX1272 (and with different boards), and it worked well.
I think the code review and some comment regarding integration with netdev is very much welcomed. Maybe @miri64 could also help on this part.

dylad commented on drivers/sx127x/sx127x.c:
```diff
@@ -0,0 +1,477 @@
+    return SX127X_INIT_OK;
+}
+
+void sx127x_init_radio_settings(sx127x_t *dev)
```
A new question again !
Why some variables are initialized with netdev and the rest with the driver functions ?

aabadie replied:
good question, I also think using netdev here is useless. Wil remove.

dylad commented on drivers/sx127x/sx127x_netdev.c:
```diff
@@ -0,0 +1,514 @@
+    uint8_t irq = dev->irq;
+    dev->irq = 0;
+
+    if (irq & SX127X_IRQ_DIO0) {
```
I think you should use a `switch`

aabadie commented on drivers/sx127x/sx127x_getset.c:
```diff
@@ -0,0 +1,826 @@
+    sx127x_reg_write(dev, SX127X_REG_FRFLSB, (uint8_t)(channel & 0xFF));
+
+    /* Restore previous operating mode */
+    sx127x_reg_write(dev, SX127X_REG_OPMODE, prev_mode);
```
indeed

aabadie commented on drivers/sx127x/sx127x_getset.c:
```diff
@@ -0,0 +1,826 @@
+    /* Disable running timers */
+    xtimer_remove(&dev->_internal.tx_timeout_timer);
+    xtimer_remove(&dev->_internal.rx_timeout_timer);
+    xtimer_remove(&dev->_internal.rx_timeout_syncword_timer);
```
Something I forgot to remove when I removed the FSK part : I started to look into this but realized I won't be able to finish it for the release.

aabadie commented on drivers/sx127x/sx127x.c:
```diff
@@ -0,0 +1,477 @@
+
+    /* Set RegOpMode value to the datasheet's default.
+       Actual default after POR is 0x09 */
+    sx127x_reg_write(dev, SX127X_REG_OPMODE, 0x00);
```
good idea! I don't understand what is 0x09, maybe a typo in the comment.

aabadie commented:
@dylad comments addressed. Thanks again for your valuable review and help.

dylad commented:
Let's have this driver merged :)

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -269,6 +269,28 @@ typedef enum {
+    /**
+     * @brief LoRA options.
+     */
+    NETOPT_LORA_BANDWIDTH,                        /**< bandwidth of LoRa */
```
Other radios have bandwidths, spreading factors, coding rates, etc., too. So think hard if all these options really need to be LoRa specific.

aabadie replied:
I wasn't aware of all those. Will update asap.

aabadie replied:
@miri64, comment addressed.

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -269,6 +269,28 @@ typedef enum {
+    NETOPT_CRC,                                   /**< CRC check enabled */
+    NETOPT_LORA_FIXED_HEADER,                     /**< Fixed header mode enabled */
+    NETOPT_LORA_IQ_INVERT,                        /**< IQ inverted enabled */
+    NETOPT_LORA_PAYLOAD_LENGTH,                   /**< Payload length */
```
What's wrong with `NETOPT_MAX_PACKET_SIZE`?

miri64 commented on drivers/sx127x/sx127x_netdev.c:
```diff
@@ -0,0 +1,513 @@
+    return 0;
+}
+
+static void _isr(netdev_t *netdev)
```
Can you please draft out: How is a receive event triggered by this function?

aabadie replied:
see [IRQ handler](https://github.com/RIOT-OS/RIOT/pull/6797/files#diff-c91e7570ffcde764db08c9803c868371R172) and the function below.

aabadie replied:
There 6 irq pins and first 4 are actually used for LoRa and FSK. The 2 remaining ones are FSK specific. The actual version of this driver doesn't support FSK

aabadie commented:
@dylad @miri64 can I squash ?

dylad commented:
All my comments are addressed or answered so ACK on my side.

aabadie commented:
I made some more tests today: the SX1272 and SX1276 modules can exchange messages with all available parameters (bandwith, spreading factors and coding rate). It works well now.

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -269,6 +269,54 @@ typedef enum {
+    NETOPT_MODEM_TYPE,
+
+    /**
+     * @brief LoRA specific options.
```
`@name` and `@{` missing, otherwise this brief will only apply to `NETOPT_LORA_SINGLE_RECEIVE`

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -269,6 +269,54 @@ typedef enum {
+    /** @brief   Radio modulation bandwidth
+     *
+     * Depends on the type of radio used and the regional regulatory and is
+     * used with LoRa/FSK modems.
```
This is more confusing then helpful. Which radio expect which type? Is the user supposed to look into the code? Also, if the type depends on the radio (even for LoRa?), how is the top-layer (e.g. a MAC layer) supposed to choose one?

aabadie replied:
> Which radio expect which type?

For the moment, only LoRa/FSK modem implement this but others could use it. I'm not enough expert with 802.15.4 PHY and implementations so I can't say.

> how is the top-layer (e.g. a MAC layer) supposed to choose one?

This is part of the LoRaMAC specifications (not done in this PR)

miri64 replied:
> This is part of the LoRaMAC specifications (not done in this PR)

Or 6LoRA some day, so is there some generic type possible?

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -269,6 +269,54 @@ typedef enum {
+
+    /** @brief   Radio spreading factor
+     *
+     * For LoRa/FSK modems, allowed values are in the range 7..12.
```
Please specify a specific type (e.g. `uint16_t`).

aabadie replied:
then it will be `uint8_t`

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -269,6 +269,54 @@ typedef enum {
+
+    /** @brief   Radio coding rate
+     *
+     * For LoRa/FSK modems, allowed values are in the range 5..8.
```
Please specify a specific type (e.g. uint16_t).

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -269,6 +269,54 @@ typedef enum {
+
+    /** @brief   Semtech module modem type.
+     *
+     * Allowed values are LoRa and FSK.
```
As string??!?

aabadie replied:
Well no of course. Will update the doc

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -269,6 +269,54 @@ typedef enum {
+     */
+    NETOPT_CODING_RATE,
+
+    /** @brief   Semtech module modem type.
```
Why is this required to be exposed?

aabadie replied:
I think it will be useful to be able to switch from one modem to other using ifconfig. This is not possible right now but it will be possible with loramac adaption.

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -269,6 +269,54 @@ typedef enum {
+    /**
+     * @brief LoRA specific options.
+     */
+    NETOPT_LORA_SINGLE_RECEIVE,      /**< When enabled, RX is turned off upon reception of a packet */
```
Doesn't sound like something that is LoRA specific either.

aabadie replied:
I don't know if other radios support this

miri64 replied:
Me neither, but they could, so there is no reason (and no harm in doing not so) to make it LoRA specific.

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -269,6 +269,54 @@ typedef enum {
+     * @brief LoRA specific options.
+     */
+    NETOPT_LORA_SINGLE_RECEIVE,      /**< When enabled, RX is turned off upon reception of a packet */
+    NETOPT_LORA_RX_TIMEOUT,          /**< Reception timeout of a LoRa packet */
```
RX timeouts can be configured for non-LoRA-radios, too.

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -269,6 +269,54 @@ typedef enum {
+     */
+    NETOPT_LORA_SINGLE_RECEIVE,      /**< When enabled, RX is turned off upon reception of a packet */
+    NETOPT_LORA_RX_TIMEOUT,          /**< Reception timeout of a LoRa packet */
+    NETOPT_LORA_TX_TIMEOUT,          /**< Transmission timeout of a LoRa packet */
```
TX timeouts can be configured for non-LoRA-radios, too.

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -269,6 +269,54 @@ typedef enum {
+    NETOPT_LORA_SINGLE_RECEIVE,      /**< When enabled, RX is turned off upon reception of a packet */
+    NETOPT_LORA_RX_TIMEOUT,          /**< Reception timeout of a LoRa packet */
+    NETOPT_LORA_TX_TIMEOUT,          /**< Transmission timeout of a LoRa packet */
+    NETOPT_LORA_SYMBOL_TIMEOUT,      /**< Symbol timeout */
```
Symbol timeouts can be configured for non-LoRA-radios, too.

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -269,6 +269,54 @@ typedef enum {
+    NETOPT_LORA_TX_TIMEOUT,          /**< Transmission timeout of a LoRa packet */
+    NETOPT_LORA_SYMBOL_TIMEOUT,      /**< Symbol timeout */
+    NETOPT_LORA_SYNCWORD,            /**< Syncword */
+    NETOPT_LORA_HOP,                 /**< Enable/disable frequency hopping */
```
Channel hoping is something other radios also support.

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -269,6 +269,54 @@ typedef enum {
+    NETOPT_LORA_RX_TIMEOUT,          /**< Reception timeout of a LoRa packet */
+    NETOPT_LORA_TX_TIMEOUT,          /**< Transmission timeout of a LoRa packet */
+    NETOPT_LORA_SYMBOL_TIMEOUT,      /**< Symbol timeout */
+    NETOPT_LORA_SYNCWORD,            /**< Syncword */
```
Does this really need to be exposed via `netdev`/`netapi`? Also: other radio types might have sync-words, too.

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -269,6 +269,54 @@ typedef enum {
+    NETOPT_LORA_SYMBOL_TIMEOUT,      /**< Symbol timeout */
+    NETOPT_LORA_SYNCWORD,            /**< Syncword */
+    NETOPT_LORA_HOP,                 /**< Enable/disable frequency hopping */
+    NETOPT_LORA_HOP_PERIOD,          /**< Frequency hopping period */
```
dito.

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -269,6 +269,54 @@ typedef enum {
+    NETOPT_LORA_SYNCWORD,            /**< Syncword */
+    NETOPT_LORA_HOP,                 /**< Enable/disable frequency hopping */
+    NETOPT_LORA_HOP_PERIOD,          /**< Frequency hopping period */
+    NETOPT_CRC,                      /**< Enable/disable CRC check */
```
This is still under the "LoRA-specific" options

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -269,6 +269,54 @@ typedef enum {
+    NETOPT_LORA_HOP_PERIOD,          /**< Frequency hopping period */
+    NETOPT_CRC,                      /**< Enable/disable CRC check */
+    NETOPT_LORA_FIXED_HEADER,        /**< Enable/disable fixed header mode */
+    NETOPT_LORA_IQ_INVERT,           /**< Enable/disable IQ inverted */
```
Doesn't seem that LoRA-specific.

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -269,6 +269,54 @@ typedef enum {
+    NETOPT_CRC,                      /**< Enable/disable CRC check */
+    NETOPT_LORA_FIXED_HEADER,        /**< Enable/disable fixed header mode */
+    NETOPT_LORA_IQ_INVERT,           /**< Enable/disable IQ inverted */
+    NETOPT_LORA_PAYLOAD_LENGTH,      /**< Payload length */
```
How does this differ from `NETOPT_MAX_PACKET_SIZE`?

aabadie replied:
imo, max packet size is a global limitation regarding hardware or physical layer. Payload length is the actual size of what was sent/received.

miri64 replied:
Actually, (at least in IEEE 802.15.4 and Ethernet) `NETOPT_MAX_PACKET_SIZE` returns the link-layer payload length (i.e. the maximum L3 packet length, wording was originally chosen since Frame == L2 PDU, Packet == L3 PDU).

aabadie replied:
Not I understand correctly.

miri64 replied:
> Payload length is the actual size of what was sent/received.

Why does this need to be returned as an option `send()` and `recv()` return this value already (including header length).

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -269,6 +269,54 @@ typedef enum {
+    NETOPT_LORA_FIXED_HEADER,        /**< Enable/disable fixed header mode */
+    NETOPT_LORA_IQ_INVERT,           /**< Enable/disable IQ inverted */
+    NETOPT_LORA_PAYLOAD_LENGTH,      /**< Payload length */
+    NETOPT_LORA_PREAMBLE_LENGTH,     /**< Preamble length */
```
Other radios have preambles, too.

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -269,6 +269,54 @@ typedef enum {
+    NETOPT_LORA_IQ_INVERT,           /**< Enable/disable IQ inverted */
+    NETOPT_LORA_PAYLOAD_LENGTH,      /**< Payload length */
+    NETOPT_LORA_PREAMBLE_LENGTH,     /**< Preamble length */
+    NETOPT_LORA_RANDOM,              /**< Random value */
```
Other radios have random values, too.

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -269,6 +269,54 @@ typedef enum {
+    NETOPT_LORA_PAYLOAD_LENGTH,      /**< Payload length */
+    NETOPT_LORA_PREAMBLE_LENGTH,     /**< Preamble length */
+    NETOPT_LORA_RANDOM,              /**< Random value */
+    NETOPT_LORA_TIME_ON_AIR,         /**< Time on air */
```
This seems like a packet-specific thing. Consider extending [`struct netdev_radio_rx_info`](http://doc.riot-os.org/structnetdev__radio__rx__info.html) instead.

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -269,6 +269,54 @@ typedef enum {
+    NETOPT_LORA_HOP,                 /**< Enable/disable frequency hopping */
+    NETOPT_LORA_HOP_PERIOD,          /**< Frequency hopping period */
+    NETOPT_CRC,                      /**< Enable/disable CRC check */
+    NETOPT_LORA_FIXED_HEADER,        /**< Enable/disable fixed header mode */
```
I can imagine other protocols might have this mode, too.

aabadie commented:
@miri64, I addressed most of your comments. I'm not convinced with IQ inverted and fixed header options, I left them as a LoRa specific one.
For payload length, I have no strong opinion.

miri64 commented:
>  I'm not convinced with IQ inverted and fixed header options, I left them as a LoRa specific one.

So there can't possibly exist any protocols that have "IQ inverted" or "fixed header" options except LoRA in your opinion? I think that's a bold claim ;-).

miri64 commented on drivers/include/sx127x.h:
```diff
@@ -0,0 +1,688 @@
+uint8_t sx127x_get_payload_length(sx127x_t *dev);
+
+/**
+ * @brief   Sets the SX127X spreading factor
```
Copy-pasta (should be "Sets the SX127X payload length")

miri64 commented on drivers/include/sx127x.h:
```diff
@@ -0,0 +1,688 @@
+bool sx127x_get_fixed_header_len_mode(sx127x_t *dev);
+
+/**
+ * @brief   Sets the SX127X spreading factor
```
Copy-pasta (should be "Sets the SX127X LoRA implicit mode" (?))

miri64 commented on drivers/include/sx127x.h:
```diff
@@ -0,0 +1,688 @@
+uint8_t sx127x_get_tx_power(sx127x_t *dev);
+
+/**
+ * @brief   Sets the SX127X spreading factor
```
Copy-pasta

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -269,6 +269,127 @@ typedef enum {
+
+    /** @brief   Get/Set the radio spreading factor
+     *
+     * For the moment, only LoRa/FSK modems support this.
```
Don't mention LoRA here. It will be a mess once we have other radios that support this.

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -269,6 +269,127 @@ typedef enum {
+
+    /** @brief   Get/Set the radio modulation bandwidth
+     *
+     * For the moment, only LoRa/FSK modems support this.
```
Don't mention LoRA here. It will be a mess once we have other radios that support this.

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -269,6 +269,127 @@ typedef enum {
+    /** @brief   Get/Set the radio modem type.
+     *
+     * Used with LoRa radio types that support both LoRa and FSK modulations.
+     * Values are passed as uint8_t and allowed values are:
```
Don't mention LoRA here. It will be a mess once we have other radios that support this.

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -269,6 +269,127 @@ typedef enum {
+
+    /** @brief   Get/Set the radio coding rate
+     *
+     * For the moment, only LoRa/FSK modems support this.
```
Don't mention LoRA here. It will be a mess once we have other radios that support this.

miri64 commented on drivers/sx127x/sx127x_netdev.c:
```diff
@@ -0,0 +1,499 @@
+            }
+
+            /* Initializes the payload size */
+            sx127x_reg_write(dev, SX127X_REG_LR_PAYLOADLENGTH, size);
```
You overwrite the payload length here anyway every time you send. Why is there a need for an `NETOPT` option then?

miri64 replied:
Also, you have a [`sx127x_set_payload_length()`](https://github.com/RIOT-OS/RIOT/pull/6797/files#diff-dada00589373226016632c37cf27ea59R671) function implemented that does effectively the same. So why not use it

miri64 commented on drivers/sx127x/sx127x_getset.c:
```diff
@@ -0,0 +1,841 @@
+    DEBUG("[DEBUG] Set payload len: %d\n", len);
+
+    if (dev->settings.lora.use_fix_len) {
+        dev->settings.lora.payload_len = len;
```
This variable get's outdated as soon as you send something. I think storing it isn't necessary.

miri64 replied:
@aabadie ping?

miri64 commented:
There are a lot of things in the code that can be made more efficient / clearer. I try to give this a proper review tomorrow so we maybe have this in the release.

aabadie commented:
> There are a lot of things in the code that can be made more efficient / clearer. I try to give this a proper review tomorrow so we maybe have this in the release.

That would be really appreciated. I'll try to address them on time... and if I can't I'll postpone feature freeze :smile:

miri64 commented:
> That would be really appreciated. I'll try to address them on time... and if I can't I'll postpone feature freeze :smile:

I think fixing #7268 has more priority than a new feature ;-) => don't postpone FF for just one feature that still needs extensive rework (potentially introducing more bugs).

aabadie commented:
> I think fixing #7268 has more priority than a new feature ;-)

Agreed ;)

> don't postpone FF for just one feature that still needs extensive rework (potentially introducing more bugs).

I don't think this PR will introduce more bugs, since it modifies very few common code (netopt.h). But I agree again that there's room for improvement in the current driver implementation

miri64 commented on drivers/sx127x/sx127x_netdev.c:
```diff
@@ -0,0 +1,499 @@
+
+        case NETOPT_PAYLOAD_LENGTH:
+            sx127x_set_payload_length(dev, *((uint8_t*) val));
+            return sizeof(uint8_t);
```
I'm less and less convinced this NETOPT is actually needed.

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -269,6 +269,114 @@ typedef enum {
+    /**
+     * @brief   Enable/disable CRC check
+     */
+    NETOPT_CRC,
```
Maybe`NETOPT_INTEGRITY_CHECK`? Not all protocols use CRC for that and this way we can use it for all those, too. Doc needs adaption, too. E.g. "Enable/disable checksum check".

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -269,6 +269,114 @@ typedef enum {
+     * * 0: FSK modem
+     * * 1: LoRa modem (default)
+     *
+     * @see drivers_sx127x
```
Dislike. Should every driver supporting this option add itself here? That's neither ideal nor generic.

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -269,6 +269,114 @@ typedef enum {
+     * * 1: 250kHz
+     * * 2: 500kHz
+     *
+     * @see drivers_sx127x
```
Please remove.

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -269,6 +269,114 @@ typedef enum {
+     * 
+     * Values are passed as uint8_t and allowed values are in the range 6..12.
+     *
+     * @see drivers_sx127x
```
Please remove.

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -269,6 +269,114 @@ typedef enum {
+
+    /** @brief   Get/Set the radio spreading factor
+     * 
+     * Values are passed as uint8_t and allowed values are in the range 6..12.
```
Please don't specify the range. You can check it in `_set()` and return `-EINVAL` for invalid values. This spreading factor range is specific to this radio.

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -269,6 +269,114 @@ typedef enum {
+     * Values are passed as uint8_t and allowed values are:
+     * * 0: 125kHz
+     * * 1: 250kHz
+     * * 2: 500kHz
```
Why not configure it in units of kHz then? Unsupported values can return `-EINVAL`.

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -269,6 +269,114 @@ typedef enum {
+     *
+     * @see drivers_sx127x
+     */
+    NETOPT_MODEM_TYPE,
```
Thinking about BLE/IEEE802.15.4 combi-devices, maybe calling it something `NETOPT_DEVICE_MODE`?

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -269,6 +269,114 @@ typedef enum {
+     *
+     * Values are passed as uint8_t and allowed values are:
+     * * 0: FSK modem
+     * * 1: LoRa modem (default)
```
I can configure my IEEE802.15.4 device to be a LoRa modem? Best thing: it is already configured as such? Oh goodie! ;-)

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -269,6 +269,114 @@ typedef enum {
+     * * 3: 4/7
+     * * 4: 4/8
+     *
+     * @see drivers_sx127x
```
Please remove

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -269,6 +269,114 @@ typedef enum {
+    /**
+     * @brief   Enable/disable fixed header mode (implicit header mode)
+     *
+     * Default is False (Explicit header mode).
```
Please remove. There is no way to ensure that for all radios.

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -269,6 +269,114 @@ typedef enum {
+    NETOPT_CODING_RATE,
+
+    /**
+     * @brief   Enable/disable fixed header mode (implicit header mode)
```
remove the `(implicit header mode)`

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -269,6 +269,114 @@ typedef enum {
+    /**
+     * @brief   Enable/disable IQ inverted
+     *
+     * Default is False.
```
Please remove. There is no way to ensure that for all radios.

miri64 commented on tests/driver_sx127x/README.md:
```diff
@@ -0,0 +1,134 @@
+
+If you have other hardware (boards, Semtech based LoRa module), you can adapt
+the configuration to your needs by updating the
+`drivers/sx127x/include/sx127x_params.h` file.
```
Better don't teach new users bad practices. They can copy the file to the application directory and then update it ;-) This leaves the original untouched and prevents PRs with changed configs ;-).

aabadie commented:
Wow, I'm wondering how I could a soooo large review... Will work on it now

miri64 commented:
> Wow, I'm wondering how I could miss a soooo large review... Will work on it now

Probably github unicorning everytime you open this PR now ^^

aabadie commented on drivers/sx127x/sx127x.c:
```diff
@@ -0,0 +1,464 @@
+    }
+}
+
+/* Following interrupt lines are not used */
```
I'd like to keep them until we add FSK support.

aabadie replied:
Finally removed them.

aabadie commented on drivers/sx127x/include/sx127x_registers.h:
```diff
@@ -0,0 +1,1738 @@
+
+#define SX127X_RF_AFCFEI_AFCAUTOCLEAR_MASK                           0xFE
+#define SX127X_RF_AFCFEI_AFCAUTOCLEAR_ON                             0x01
+#define SX127X_RF_AFCFEI_AFCAUTOCLEAR_OFF                            0x00  /**< Defaul
```
This value is related to FSK

aabadie commented on drivers/sx127x/include/sx127x_registers.h:
```diff
@@ -0,0 +1,1738 @@
+ */
+
+/** Common settings */
+#define SX127X_REG_LR_OPMODE                                         0x01
```
This is the worst comment :)

aabadie commented on drivers/sx127x/sx127x_netdev.c:
```diff
@@ -0,0 +1,499 @@
+        sx127x_on_dio5(dev);
+        break;
+    default:
+        break;
```
according to RIOT coding style, that points to [Linux kernel coding style guidelines](https://www.kernel.org/doc/Documentation/process/coding-style.rst), this one is correct, the others aren't.

aabadie commented on drivers/include/sx127x.h:
```diff
@@ -0,0 +1,688 @@
+    bool freq_hop_on;                                           /**< Enable frequency hopping */
+    uint8_t freq_hop_period;                                    /**< Frequency hop period */
+    bool iq_inverted;                                           /**< Set inverted IQ */
+    bool rx_continuous;                                         /**< Use continuous reception */
```
> There are other bools in this struct that can all be summarized to a flag field, saving tons of bytes

I know, but I didn't want to get into optimizations first and have a clean and working driver implementation

miri64 commented:
2017-06-29 17:58 GMT+02:00 Alexandre Abadie <notifications@github.com>:

> *@aabadie* commented on this pull request.
> ------------------------------
>
> In drivers/sx127x/sx127x_netdev.c
> <https://github.com/RIOT-OS/RIOT/pull/6797#discussion_r124841943>:
>
> > +        sx127x_on_dio1(dev);
> +        break;
> +    case SX127X_IRQ_DIO2:
> +        sx127x_on_dio2(dev);
> +        break;
> +    case SX127X_IRQ_DIO3:
> +        sx127x_on_dio3(dev);
> +        break;
> +    case SX127X_IRQ_DIO4:
> +        sx127x_on_dio4(dev);
> +        break;
> +    case SX127X_IRQ_DIO5:
> +        sx127x_on_dio5(dev);
> +        break;
> +    default:
> +        break;
>
> according to RIOT coding style, that points to Linux kernel coding style
> guidelines
> <https://www.kernel.org/doc/Documentation/process/coding-style.rst>, this
> one is correct, the others aren't.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/RIOT-OS/RIOT/pull/6797#discussion_r124841943>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AApPPHq3svc2A2UJBcnaMDtbZuqymVmiks5sI8mWgaJpZM4MqR_O>
> .
>
Mhhhh, but a quick `git grep -A 2 -e "\<switch\> (.*) {"` shows, that we
definitely prefer the other style (and afaik that is what is generated by
`uncrustify`.

aabadie commented on drivers/sx127x/sx127x.c:
```diff
@@ -0,0 +1,464 @@
+    _init_timers(dev);
+    xtimer_usleep(1000); /* wait 1 millisecond */
+
+    sx127x_reset(dev);
```
I removed the first one. Doesn't have any effect.

aabadie commented:
It's becoming very difficult to open this PR on Github.

I would like to create another one. Is it ok for you @miri64 ?

miri64 commented:
> It's becoming very difficult to open this PR on Github.

Agreed.

> I would like to create another one. Is it ok for you @miri64 ?

I fear I would loose track of my review due to that :-/.

aabadie commented:
It's becoming very difficult to open this PR on Github.

Agreed.

I would like to create another one. Is it ok for you @miri64 ?

I fear I would loose track of my review due to that :-/.I know, that's why I ask ;)

miri64 commented:
> I know, that's why I ask ;)

So let's force ourselves staring at unicorns then :grin:

jia200x commented on drivers/sx127x/sx127x_netdev.c:
```diff
@@ -0,0 +1,499 @@
+                                   SX127X_RF_LORA_INVERTIQ_TX_MASK & SX127X_RF_LORA_INVERTIQ_RX_MASK) |
+                                  SX127X_RF_LORA_INVERTIQ_RX_OFF | SX127X_RF_LORA_INVERTIQ_TX_ON));
+                sx127x_reg_write(dev, SX127X_REG_LR_INVERTIQ2,
+                                 SX127X_RF_LORA_INVERTIQ2_ON);
```
The SX127X_REG_LR_INVERTIQ2 can be inside the sx127x_set_iq_invert function.

BTW, this one [is not documented in the datasheet](https://github.com/4refr0nt/lora-gw/blob/master/sx1276/sx1276.cxx#L1327). Seems to be an undocumented magic number according to that :P

aabadie commented:
@miri64, I addressed your comments.
In the meantime, I also applied the new const policy here as well and trusted @jia200x with the IQ invert setter. I'll test tomorrow if this change it doesn't introduce a regression.

aabadie commented:
Let me know when I can squash :)

miri64 commented on drivers/sx127x/sx127x_netdev.c:
```diff
@@ -0,0 +1,541 @@
+            /* todo */
+            break;
+        case SX127X_MODEM_LORA:
+            sx127x_set_iq_invert(dev, dev->settings.lora.iq_inverted);
```
I might be mistaken, but isn't this field set in `sx127x_set_iq_invert()`, implying that the device (and its registers) has this option already set accordingly? If the registers can change due to side effects, I would prefer that the setting of the field (via the NETOPT and on initialization) is done in a separate function from setting the registers.

aabadie replied:
totally removed

miri64 commented on drivers/sx127x/sx127x_getset.c:
```diff
@@ -0,0 +1,856 @@
+                                   & SX127X_RF_LORA_INVERTIQ_TX_MASK & SX127X_RF_LORA_INVERTIQ_RX_MASK)
+                                  | SX127X_RF_LORA_INVERTIQ_RX_OFF | SX127X_RF_LORA_INVERTIQ_TX_OFF));
+                sx127x_reg_write(dev, SX127X_REG_LR_INVERTIQ2, SX127X_RF_LORA_INVERTIQ2_OFF);
+            }
```
Can't this block get the same treatment as in `_send()`?

miri64 commented on drivers/include/sx127x.h:
```diff
@@ -0,0 +1,704 @@
+    uint8_t power;                     /**< Signal power */
+    uint8_t bandwidth;                 /**< Signal bandwidth */
+    uint8_t datarate;                  /**< Spreading factor rate, e.g datarate */
+    bool low_datarate_optimize;        /**< Optimize for low data rate */
```
What about these bools? Also this and the other structs could benefit of some [reordering](http://www.catb.org/esr/structure-packing/)

aabadie replied:
After discussing this with @adjih and @astralien3000, this is very low priority : the structures are not duplicated in memory so the number of saved bytes is very low as the cost of more flash used and slower execution. I prefer to leave as it is now.

miri64 replied:
I'm not sure I agree.

First of all there is the number of saved memory bytes, which *is* significant in my opinion. Let's look at the current case and an optimized case in comparison. In the current case the memory on a 32-bit platform looks like this

```
0                   1                   2                   3
0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2
+---------------+---------------+---------------+---------------+
| power         | bandwidth     | datarate      |               |   32 bytes
+---------------+---------------+---------------+---------------+
| low_datarate_optimize                                         |   64 bytes
+---------------+---------------+-------------------------------+
| coderate      |               | preamble_len                  |   96 bytes
+---------------+---------------+-------------------------------+
| use_fix_len                                                   |  128 bytes
+---------------------------------------------------------------+
| crc_on                                                        |  160 bytes
+---------------------------------------------------------------+
| freq_hop_on                                                   |  192 bytes
+---------------------------------------------------------------+
| freq_hop_per  |                                               |  224 bytes
+---------------+-----------------------------------------------+
| iq_inverted                                                   |  256 bytes
+---------------------------------------------------------------+
| rx_continuous                                                 |  288 bytes
+---------------------------------------------------------------+
| rx_timeout                                                    |  320 bytes
+---------------------------------------------------------------+
| tx_timeout                                                    |  352 bytes
+---------------------------------------------------------------+
```

Compare that to a reordered struct with flags instead of bools:

```
0                   1                   2                   3
0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2
+---------------+---------------+---------------+---------------+
| power         | bandwidth     | datarate      | coderate      |   32 bytes
+---------------+---------------+-------------------------------+
| preamble_len                  | freq_hop_on   | flags         |   64 bytes
+-------------------------------+-------------------------------+
| rx_timeout                                                    |   96 bytes
+---------------------------------------------------------------+
| tx_timeout                                                    |  128 bytes
+---------------------------------------------------------------+
```

That is a gain of 224 bytes, which is huge in our terms (remember, we support systems that only have kBytes of RAM). Even in the non-reordered case we would still gain 196 bytes! If that doesn't drive my point home: Even on 16-bit systems (and even 8-bit platforms maybe, I don't know exactly how they pack their structs tbh), where we typically have even less RAM we gain 160 bytes with that optimization.

Regarding the flash and execution time we have to distinguish between setting (`x = true` vs `x |= FLAG`), unsetting (`x = false` vs `x &= ~FLAG`) and comparing (`(x)` vs `(x & FLAG)`) the variable/flag since they differ in the flag case 

* For setting the byte we have the same number of ops on ARM. `MOV` for the bool case, `ORR` for the flag case.
* For unsetting ARM seems to have a special operation for bitwise unset (`BICS` [?] at least according to my analysis) so we again have have one op vs one (bool uses MOV again)
* Finally I think we win again some ROM when using flags in the compare case, since the compiler translates the `(x)` check to an `UXTB` and `CMP`, while (x & FLAG) is compiled to a simple bitwise and + compare register update `ANDS`.

All in all, especially since you think the priority for that is low I would argue that we then should do it right the first time, otherwise we run around with unoptimized code for years ;-).

miri64 replied:
Oops, divide all the byte numbers I gave by 8.... Those were bits **facepalm**. Is still think this is beneficial

miri64 commented on drivers/sx127x/sx127x.c:
```diff
@@ -0,0 +1,419 @@
+    sx127x_set_crc(dev, SX127X_PAYLOAD_CRC_ON);
+    sx127x_set_symbol_timeout(dev, SX127X_SYMBOL_TIMEOUT);
+    sx127x_set_preamble_length(dev, SX127X_PREAMBLE_LENGTH);
+    // sx127x_set_payload_length(dev, SX127X_PAYLOAD_LENGTH);
```
Please remove commented out code.

miri64 replied:
(and use c-style comments, if any :P)

aabadie replied:
or uncomment...

miri64 commented on sys/include/net/netopt.h:
```diff
@@ -269,10 +269,85 @@ typedef enum {
+    /**
+     * @brief   Set payload length as uint8_t.
+     */
+    NETOPT_PAYLOAD_LENGTH,
```
Still not convinced this `NETOPT` is necessary at all, but maybe @jia200x or @dylad can shed some light on this regarding LoRAMAC.

jia200x replied:
This NETOPT tells the driver the maximum payload length permitted. The ongoing GNRC LoRaWAN implementation sets this to 0 (Unlimited Packet Length). But unless we hardcode this in the driver, I vote to leave it as is.

miri64 replied:
Now that the page isn't unicorning anymore, for better overview I post it here again:

That is what `NETOPT_MAX_PACKET_SIZE` is for (and used for: it calls `set_max_payload_len()`). This NETOPT, when set calls
the devices `set_payload_len()` function, which is also called every time
a packet is sent in the netdev driver's `_send` function (effectively changing the value).

aabadie replied:
I removed it since I'm now also convinced it useless in the current state. It can be added later if really required for LoRaWAN and `MAX_PACKET_SIZE` doesn't fit.

jia200x commented:
@aabadie @miri64 What happened to the NETOPT_LORA_SYNCWORD option? It's needed for LoRaWAN. Public and private LoRaWAN network differ on this value (0x39 vs 0x34) , and should be set by the MAC layer.

aabadie commented:
> What happened to the NETOPT_LORA_SYNCWORD option? It's needed for LoRaWAN. Public and private LoRaWAN network differ on this value (0x39 vs 0x34) , and should be set by the MAC layer.

I know and I removed it because LoraWAN is not part of this PR.

RIOT-notifier commented:
>
> In sys/include/net/netopt.h
> <https://github.com/RIOT-OS/RIOT/pull/6797#discussion_r124946004>:
>
> > +    NETOPT_CODING_RATE,
> +
> +    /**
> +     * @brief   Enable/disable fixed header mode.
> +     */
> +    NETOPT_FIXED_HEADER,
> +
> +    /**
> +     * @brief   Enable/disable IQ inverted.
> +     */
> +    NETOPT_IQ_INVERT,
> +
> +    /**
> +     * @brief   Set payload length as uint8_t.
> +     */
> +    NETOPT_PAYLOAD_LENGTH,
>
> This NETOPT tells the driver the maximum payload length permitted. The
> ongoing GNRC LoRaWAN implementation sets this to 0 (Unlimited Packet
> Length). But unless we hardcode this in the driver, I vote to leave it as
> is.
>

That is what `NETOPT_MAX_PACKET_SIZE` is for. This NETOPT, when set calls
the devices `set_payload_len()` function,  which is also called every time
a packet is sent in the netdev driver's `_send` function.

miri64 commented:
(that was me ^^")

miri64 commented:
>> What happened to the NETOPT_LORA_SYNCWORD option? It's needed for LoRaWAN. Public and private LoRaWAN network differ on this value (0x39 vs 0x34) , and should be set by the MAC layer.
>
> I know and I removed it because LoraWAN is not part of this PR.

I did not ask for that. And I still would prefer a `NETOPT_SYNCWORD` ;-)

aabadie commented:
@miri64 comment addressed, except the struct reordering and bool flags, see comment above.

aabadie commented:
@miri64, I pushed a couple of commits that fix typo in readme and doxygen.

miri64 commented on drivers/include/sx127x.h:
```diff
@@ -0,0 +1,703 @@
+typedef struct {
+    /* Data that will be passed to events handler in application */
+    uint32_t last_channel;             /**< Last channel in frequency hopping sequence */
+    bool is_last_cad_success;          /**< Sign of success of last CAD operation (activity detected) */
```
This bool can stay, since the we can not win any RAM by re-packing. The two xtimers should go to the top of the struct however.

miri64 commented on drivers/include/sx127x.h:
```diff
@@ -0,0 +1,703 @@
+    uint8_t state;                     /**< Radio state */
+    uint8_t modem;                     /**< Driver model (FSK or LoRa) */
+    uint32_t channel;                  /**< Radio channel */
+    sx127x_lora_settings_t lora;       /**< LoRa settings */
```
Please move to the top of the struct

miri64 commented on drivers/include/sx127x.h:
```diff
@@ -0,0 +1,703 @@
+ */
+typedef struct {
+    uint8_t state;                     /**< Radio state */
+    uint8_t modem;                     /**< Driver model (FSK or LoRa) */
```
Please move these two to the bottom of the struct.

miri64 commented:
Apart from the optimization stuff above I'm fine with the PR btw. If that is solved we can merge IMHO

miri64 commented:
(maybe someone else than @aabadie should also test, since I don't have the hardware).

dylad commented:
@miri64 You're lucky, I'm about to run some test with custom hardware on the current state. I'll share results after that.

miri64 commented:
> @miri64 You're lucky, I'm about to run some test with custom hardware on the current state. I'll share results after that.

Well I would at least wait until @aabadie commited the proposed optimization, since there might be some risk of introducing now bugs ;D.

dylad commented:
Then I'll wait for @aabadie commits. But I have to test it right now for company stuff ;)

miri64 commented on drivers/include/sx127x.h:
```diff
@@ -0,0 +1,704 @@
+/**
+ * @brief   Generates 32 bits random value based on the RSSI readings
+ *
+ * @pre     This function sets the radio in LoRa mode and disables all
```
This is not a *pre*-condition, since it describes steps to be made after the function call. It's not even a postcondition (would be signified with `@post`), which would describe the state after execution. Best use `@attention` here.

aabadie commented:
Bit optimization done and pushed. I tested on hardware and it still works (hopefully).

Concerning memory save. Here is the size info before this change:
```
  text	   data	    bss	    dec	    hex	filename
  24752	    500	   4108	  29360	   72b0
```

And now:
```
   text	   data	    bss	    dec	    hex	filename
  24832	    500	   4100	  29432	   72f8
```

I'm loosing 82 bytes in ROM but have 8 bytes less in RAM. And I tested the combination you proposed but none of them do better.

dylad commented on drivers/sx127x/sx127x.c:
```diff
@@ -0,0 +1,419 @@
+            }
+            break;
+        default:
+            puts("sx127x_on_dio0: Unknown state");
```
@aabadie I encountered a potential bug : 
I sent 7 bytes "123456\0" using tests/driver_sx127x and the send command
If I keep the touch enter pressed (yes I'm a monster) It works so far.
Then I re-press the enter key and I get
sx127x_on_dio0: Unknown state
Note: data are emitted ! But there is no more Transmission completed message anymore (no tx done interrupt received ? Weird...)
I'm now wondering : If this case happens in reception. No more RX done interrupt no more message received -> we're stuck.
Note : if I get sx127x_on_dio0: Unknown state, then call listen command and try to re-sent data after that. Rx done interrupt comes back.
There might be something to do...
I'm now wondering :

miri64 commented on drivers/include/sx127x.h:
```diff
@@ -538,21 +545,21 @@ uint8_t sx127x_get_coding_rate(const sx127x_t *dev);
 
 /**
- * @brief   Gets the SX127X LoRa RX single mode
+ * @brief   Check is the SX127X LoRa RX single mode is enabled/disabled
```
Check if

miri64 replied:
Ping @aabadie?

aabadie commented:
Tested on ARM STM32F4 (nucleo-f411).

miri64 commented on drivers/include/sx127x.h:
```diff
@@ -192,21 +192,28 @@ enum {
+#define SX127X_ENABLE_CRC_FLAG_MASK                  (1 << 2)
+#define SX127X_CHANNEL_HOPPING_FLAG_MASK             (1 << 3)
+#define SX127X_IQ_INVERTED_FLAG_MASK                 (1 << 4)
+#define SX127X_RX_CONTINUOUS_FLAG_MASK               (1 << 5)
```
These are not masks, they are flags. So don't call them that. Also: can you try if setting the flags explicitly (i.e. `0x01`, `0x02`, `0x04`, etc) makes a different in the code size?

aabadie replied:
it is worse

miri64 replied:
That is surprising oO, but okay

miri64 commented on drivers/include/sx127x.h:
```diff
@@ -192,21 +192,28 @@ enum {
-    bool rx_continuous;                /**< Use continuous reception */
+    uint8_t flags;                     /**< Boolean flags */
     uint32_t rx_timeout;               /**< RX timeout in symbols */
     uint32_t tx_timeout;               /**< TX timeout in symbols */
```
Can you try if moving the two 32-bit numbers to the struct makes a different RAM-wise.

aabadie replied:
to which struct ? 
In this one, it has no effect.

miri64 replied:
Yes and okay.

aabadie commented:
@dylad thanks for testing. I can reproduce what you describe and to me this is a problem with the state machine: we are trying to send while the radio is transmitting. And we end up in the radio being in the wrong state when TX is done or CAD is done (both event can trigger DIO interrupt).

aabadie commented:
The wrong status is IDLE @dylad. Should not happen, still digging

aabadie commented:
@dylad in fact what you are trying to do is not legal: you are flooding the channel whereas you can only have access to it 1% of time legally.
I have a fix: prevent send if radio is already in send mode. This is more in the spirit of LoRa I think. And it prevents the radio from being in a bad state.

aabadie commented:
@dylad fix pushed, @miri64 remaining comments addressed.

Retested on hardware: all good now.

Can I squash ?

miri64 commented on drivers/include/sx127x.h:
```diff
@@ -545,7 +545,7 @@ uint8_t sx127x_get_coding_rate(const sx127x_t *dev);
 
 /**
- * @brief   Check is the SX127X LoRa RX single mode is enabled/disabled
+ * @brief   Check if the SX127X LoRa RX single mode is enabled/disabled
```
"Check*s*"

aabadie commented:
remaining nit fixed. @dylad please ACK :)

miri64 commented:
@aabadie in the meantime you may squash :-).

aabadie commented:
squashed

dylad commented:
`in fact what you are trying to do is not legal`
@aabadie you're perfectly right and I know the EU regulation but I'm testing ;)
To tell you the truth, I'm currently digging into my code to spot a **_potential_** LoRa rx issue which kills reception.
I'm tracking a bug that I'm unable to reproduce it at my company place but it seldom happens at my client's place. We're still not sure where the bug is located but I'm investigating. I would like to flood in reception one of our nodes to simulate a dense urban area using radio to check if I can create possible issue on reception.

Note : our bug may not be related to RIOT or even to software but nevermind I had to track it and patch it.

dylad commented:
@aabadie @miri64 
Sorry I cannot re-test the last commit right now. Best I can do is monday evening.
My last test was fine and I like the recent changes.
ACK now or monday, what do you prefer ? :)

aabadie commented:
ACK now or monday, what do you prefer ?Now ;)Otherwise, we'll have to postpone and it won't be merged before mid-July at best.I'll announce feature freeze later today or during the week-end.By the way, we retested again with @adjih and it was all good.

miri64 commented:
@aabadie then let @adjih ACK ;-)

aabadie commented:
@aabadie then let @adjih ACK ;-)


  
  




I think I'll postpone this one finally...

miri64 commented:
> I think I'll postpone this one finally...

Why is @adjih already gone? Then I'll proxy-ACK for both @adjih and @dylad... Two witnesses except you of a working device make me confident enough.

dylad commented:
If we can merged now let's do this ! ACK on my side.

miri64 commented:
:tada: We made it

dylad commented:
Great news ! It was a very long way to go ;)
BTW, you can safely close #6002 (dunno for #6645)

aabadie commented:
Can't believe it's merged :)

Many thanks to all people involved. Next step is LoraWAN adaption.

aabadie commented:
:beers:

dylad commented:
`Next step is LoraWAN adaption.` :+1: 
Let's do it !

emmanuelsearch commented:
Congrats to all involved!

miri64 commented on drivers/sx127x/include/sx127x_netdev.h:
```diff
@@ -0,0 +1,48 @@
+ */
+typedef struct netdev_radio_lora_packet_info {
+    uint8_t rssi;           /**< RSSI of a received packet */
+    uint8_t lqi;            /**< LQI of a received packet */
```
I just had a look at this code again due to an unrelated issue I discussed with @kYc0o, so sorry for the "late review" ;-).

In the [`recv()` function of this driver](https://github.com/RIOT-OS/RIOT/pull/6797/files#diff-37ee9c4f2a20ec7a38d429c04255744aR155) it says that LoRA has no LQI, so why is it in this struct?

jia200x replied:
indeed it shouldn't be here. I will fix it, since I require it for #11022
